### PR TITLE
upgrade RocksDB version to 9.4

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Implements https://arangodb.atlassian.net/browse/APM-831
Enterprise companion PR: https://github.com/arangodb/enterprise/pull/1090

Upgrade bundled version of RocksDB to 9.4
Note: never mind the branch name. This PR was created in September 2022 when RocksDB 7.7 was the most recent RocksDB version. We upgraded the PR a few times since then, but kept the original branch names.

The default values for the following RocksDB-related startup options were adjusted with this PR: 

- `--rocksdb.compaction-read-ahead-size` was changed from 2MB to 8MB.
- `--rocksdb.max-subcompactions` was changed from 2 to 4. 
- `--rocksdb.max-total-wal-size` was changed from 80MB to 256MB.
- `--rocksdb.throttle-slots` was changed from 120 to 128.
- `--rocksdb.throttle-scaling-factor` was changed from 17 to 192.
- `--rocksdb.pending-compactions-slowdown-trigger` was changed from 128KB to 1GB.
- `--rocksdb.pending-compactions-stop-trigger` was changed from 16GB to 32GB. 
- `--rocksdb.recycle-log-file-num` was changed from 0 to 4. 
- `--rocksdb.partition-files-for-documents` was changed from false to true. 

The startup option `--rocksdb.throttle-slow-down-writes-trigger` is obsoleted by this PR.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -

#### Related Information

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1090
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/APM-831
- [ ] Design document:

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 1]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
                       arangodb::options::Flags::OnSingle))
       .setIntroducedIn(31000);
 
+  options
+      ->addOption("--rocksdb.enable-blob-files",
+                  "enable blob files for documents column family",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:974]:
```suggestion
                  "Enable blob files for the documents column family.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 2]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+                      arangodb::options::Flags::DefaultNoComponents,
+                      arangodb::options::Flags::OnAgent,
+                      arangodb::options::Flags::OnDBServer,
+                      arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption(
+          "--rocksdb.enable-blob-cache",
+          "enable caching of blobs in block cache for documents column family",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:987]:
```suggestion
          "Enable caching of blobs in the block cache for the documents "
          "column family.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 3]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+              arangodb::options::Flags::OnAgent,
+              arangodb::options::Flags::OnDBServer,
+              arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption(
+          "--rocksdb.min-blob-size",
+          "minimum size threshold for storing blobs in the documents column "
+          "family (in bytes, requires `--rocks.enable-blob-files`)",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1001]:
```suggestion
          "The minimum size threshold for storing blobs in the documents "
          "column family (in bytes). Requires `--rocks.enable-blob-files`.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 4]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+              arangodb::options::Flags::DefaultNoComponents,
+              arangodb::options::Flags::OnAgent,
+              arangodb::options::Flags::OnDBServer,
+              arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption("--rocksdb.blob-file-size",
+                  "size limit for blob files in the documents column family "
+                  "(in bytes, requires `--rocksdb.enable-blob-files`)",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1014]:
```suggestion
                  "The size limit for blob files in the documents column "
                  "family (in bytes). Requires `--rocksdb.enable-blob-files`.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 5]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+                      arangodb::options::Flags::DefaultNoComponents,
+                      arangodb::options::Flags::OnAgent,
+                      arangodb::options::Flags::OnDBServer,
+                      arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption("--rocksdb.blob-file-starting-level",
+                  "level from which on to use blob files in the documents "
+                  "column family (requires `--rocksdb.enable-blob-files`)",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1027]:
```suggestion
                  "The level from which on to use blob files in the documents "
                  "column family. Requires `--rocksdb.enable-blob-files`.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 6]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+                      arangodb::options::Flags::OnAgent,
+                      arangodb::options::Flags::OnDBServer,
+                      arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption(
+          "--rocksdb.blob-garbage-collection-age-cutoff",
+          "age cutoff for garbage collecting blobs in the documents column "
+          "family (requires `--rocksdb.enable-blob-files`)",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1041]:
```suggestion
          "The age cutoff for garbage collecting blobs in the documents "
          "column family. Requires `--rocksdb.enable-blob-files`.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 7]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+              arangodb::options::Flags::OnAgent,
+              arangodb::options::Flags::OnDBServer,
+              arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption("--rocksdb.blob-garbage-collection-force-threshold",
+                  "garbage threshold for scheduling targeted compactions for "
+                  "oldest blob files in the documents column family (requires "
+                  "`--rocksdb.enable-blob-files`)",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1055]:
```suggestion
                  "The garbage threshold for scheduling targeted compactions "
                  "for the oldest blob files in the documents column family. "
                  "Requires `--rocksdb.enable-blob-files`.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 8]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+                      arangodb::options::Flags::OnAgent,
+                      arangodb::options::Flags::OnDBServer,
+                      arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  TRI_ASSERT(::compressionTypes.contains(_blobCompressionType));
+  options
+      ->addOption("--rocksdb.blob-compression-type",
+                  "compression algorithm to use for blobs within the documents "
+                  "column family (requires `--rocksdb.enable-blob-files`)",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1070]:
```suggestion
                  "The compression algorithm to use for blobs within the "
                  "documents column family. "
                  "Requires `--rocksdb.enable-blob-files`.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 9]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+                      arangodb::options::Flags::OnAgent,
+                      arangodb::options::Flags::OnDBServer,
+                      arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption(
+          "--rocksdb.enable-blob-garbage-collection",
+          "enable blob garbage compaction during compaction in the documents "
+          "column family (requires `--rocksdb.enable-blob-files`)",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1085]:
```suggestion
          "Enable blob garbage compaction during compaction in the "
          "documents column family. Requires `--rocksdb.enable-blob-files`.",
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 10]
```diff
@@ -928,6 +969,142 @@ void RocksDBOptionFeature::collectOptions(
...
+              arangodb::options::Flags::Experimental,
+              arangodb::options::Flags::DefaultNoComponents,
+              arangodb::options::Flags::OnAgent,
+              arangodb::options::Flags::OnDBServer,
+              arangodb::options::Flags::OnSingle))
+      .setIntroducedIn(31100);
+
+  options
+      ->addOption("--rocksdb.prepopulate-blob-cache",
+                  "prepopulate blob cache on flushes",
```
Comment (Simran-B) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1097]:
```suggestion
                  "Prepopulate the blob cache on flushes.",
```

**Review by Simran-B — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 11]
```diff
@@ -194,10 +194,6 @@ endif ()
 set(ENTERPRISE_INCLUDE_DIR "enterprise")
 option(USE_ENTERPRISE "enable enterprise build" OFF)
 
-# we want the following definitions to be in effect for both rocksdb and arangodb
-add_definitions("-DNROCKSDB_THREAD_STATUS")
-add_definitions("-DROCKSDB_SUPPORT_THREAD_LOCAL")
```
Comment (jsteemann) [inline CMakeLists.txt:199]: note: the new version of RocksDB does not react to a define `ROCKSDB_SUPPORT_THREAD_LOCAL` anymore, so the line can be removed.

File: CMakeLists.txt [hunk 12]
```diff
@@ -194,10 +194,6 @@ endif ()
 set(ENTERPRISE_INCLUDE_DIR "enterprise")
 option(USE_ENTERPRISE "enable enterprise build" OFF)
 
-# we want the following definitions to be in effect for both rocksdb and arangodb
-add_definitions("-DNROCKSDB_THREAD_STATUS")
```
Comment (jsteemann) [inline CMakeLists.txt:198]: note: the new version of RocksDB still checks for a define `NROCKSDB_THREAD_STATUS`, but if it is set, then it does not link anymore. So it seems that the option is now unsupported and we shouldn't set it anymore. Thus removed it here.

File: arangod/RocksDBEngine/RocksDBRateLimiterThread.cpp [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+              // decreate write rate. for this use the original scaling factor.
+              return oldRate - (oldRate - targetRate) / _scalingFactor;
+            }
+          });
+
+          // write rate must always be between configured minimum and maximum
+          // write rates.
+          newRate = std::clamp(newRate, _minWriteRate, _maxWriteRate);
+
+          LOG_TOPIC("37e36", INFO, Logger::ENGINES)
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBRateLimiterThread.cpp:267]: Note to self: log level needs to be lowered to `DEBUG` here. The `INFO` is currently only there to have enough output during some performance test runs.

**Pull request closed without merging**